akka-javasdk/src/main/scala/akka/javasdk/impl/agent/AgentImpl.scala

```diff
@@ -219,17 +222,33 @@ private[impl] final class AgentImpl[A <: Agent](
     }
   }
 
-  override def transformResponse(modelResponse: String, responseType: Class[_]): BytesPayload = {
-    if (responseType == classOf[String]) {
-      serializer.toBytes(modelResponse)
-    } else {
-      // We might be able to bypass serialization roundtrip here, but might be good to catch invalid json
-      // as early as possible.
-      // The content type isn't used in this fromBytes.
-      val obj = serializer.fromBytes(
-        responseType,
-        new BytesPayload(ByteString.fromString(modelResponse), JsonSerializer.JsonContentTypePrefix + "object"))
-      serializer.toBytes(obj)
+  override def transformResponse(
+      modelResponse: String,
+      responseType: Class[_],
+      mappingFunction: Option[Any => Any],
+      failureMapping: Option[Throwable => Any]): BytesPayload = {
+    try {
+      if (responseType == classOf[String]) {
+        val mappedObj = mappingFunction.map(_.apply(modelResponse)).getOrElse(modelResponse)
+        serializer.toBytes(mappedObj)
+      } else {
+        // We might be able to bypass serialization roundtrip here, but might be good to catch invalid json
+        // as early as possible.
+        // The content type isn't used in this fromBytes.
+
+        val obj = serializer.fromBytes(
+          responseType,
+          new BytesPayload(ByteString.fromString(modelResponse), JsonSerializer.JsonContentTypePrefix + "object"))
+
+        val mappedObj = mappingFunction.map(_.apply(obj)).getOrElse(obj)
+
+        serializer.toBytes(mappedObj)
+      }
+    } catch {
+      case e: IllegalArgumentException if failureMapping.isDefined =>
+        //trying to recover
+        val any = failureMapping.get.apply(new JsonParsingException(e.getMessage, e, modelResponse))
```

Comment:
I think we should call that method also directly from the runtime, in the `kalix.runtime.agent.Agent#requestModel` method.

Reply:
but that requires more discussion, which errors should be recoverable